### PR TITLE
[SPARK-26870][SQL] Move to_avro/from_avro into functions object due to Java compatibility

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, looks, to use `from_avro` and `to_avro` in Java APIs side,

```java
import static org.apache.spark.sql.avro.package$.MODULE$;

MODULE$.to_avro
MODULE$.from_avro
```

This PR targets to deprecate and move both functions under `avro` package into `functions` object like the way of our `org.apache.spark.sql.functions`.

Therefore, Java side can import:

```java
import static org.apache.spark.sql.avro.functions.*;
```

and Scala side can import:

```scala
import org.apache.spark.sql.avro.functions._
```

## How was this patch tested?

Manually tested, and unit tests for Java APIs were added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @coud-fan, @gengliangwang, @gaborgsomogyi

File: external/avro/src/main/scala/org/apache/spark/sql/avro/functions.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.Column
+
+
+// scalastyle:off
+object functions {
+// scalastyle:on
+
+  /**
```
Comment (HyukjinKwon) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/functions.scala]: below codes were simply copied and pasted.

File: external/avro/src/test/java/org/apache/spark/sql/avro/JavaAvroFunctionsSuite.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import static org.apache.spark.sql.avro.functions.from_avro;
+
+
+public class JavaAvroFunctionsSuite {
+  private transient TestSparkSession spark;
+
+  @Before
+  public void setUp() {
+    spark = new TestSparkSession();
+    spark.loadTestData();
```
Comment (gaborgsomogyi) [inline external/avro/src/test/java/org/apache/spark/sql/avro/JavaAvroFunctionsSuite.java]: Is it really needed?

File: external/avro/src/test/java/org/apache/spark/sql/avro/JavaAvroFunctionsSuite.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  @After
+  public void tearDown() {
+    spark.stop();
+  }
+
+  private static void checkAnswer(Dataset<Row> actual, Dataset<Row> expected) {
+    String errorMessage = QueryTest$.MODULE$.checkAnswer(
```
Comment (gaborgsomogyi) [inline external/avro/src/test/java/org/apache/spark/sql/avro/JavaAvroFunctionsSuite.java]: Nit: No line break necessary.

File: docs/sql-data-sources-avro.md [hunk 4]
```diff
@@ -109,7 +109,8 @@ val query = output
 </div>
 <div data-lang="java" markdown="1">
 {% highlight java %}
-import org.apache.spark.sql.avro.*;
+import static org.apache.spark.sql.functions.col;
```
Comment (gaborgsomogyi) [inline docs/sql-data-sources-avro.md]: Wondering why `col` needed?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala [hunk 5]
```diff
@@ -52,12 +50,12 @@ package object avro {
    * @since 3.0.0
    */
   @Experimental
+  @deprecated("Please use 'org.apache.spark.sql.avro.functions.from_avro' instead.", "3.0.0")
```
Comment (viirya) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: This is since 3.0.0. So I think we don't need to deprecate it but just remove it?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/functions.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.Column
+
+
+// scalastyle:off
+object functions {
+// scalastyle:on
```
Comment (viirya) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/functions.scala]: What this scalastyle off is for?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala (hunk 5 above)
Comment (gaborgsomogyi) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: It's added in 2.4 and deprecated in 3.0. Why remove it?

Comment (viirya) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: There is `@since 3.0.0` tag. I think it means this API is added in 3.0.0.

File: docs/sql-data-sources-avro.md (hunk 4 above)
Comment (HyukjinKwon) [reply docs/sql-data-sources-avro.md]: Because it's used below.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/functions.scala (hunk 6 above)
Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/functions.scala]: This one, `ObjectNamesChecker`. I just copied this from `functions.scala`. Let me specify the rule id.

**Comment by HyukjinKwon** [comment]
retest this please

File: external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala (hunk 5 above)
Comment (gaborgsomogyi) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: That's a question from my side because I would read `deprecated since 3.0`. If you're 100% sure then I've learned something.

Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: It's correct to remove it out. Thanks for pointing this out.

Comment (gaborgsomogyi) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: I've just checked scala source:
```
 *  @param  since   a string identifying the first version in which the definition was deprecated
```

Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: 3.0.0 is not released. Therefore, technically we don't have the first version.

File: docs/sql-data-sources-avro.md (hunk 4 above)
Comment (gaborgsomogyi) [reply docs/sql-data-sources-avro.md]: Now I see, it's a fix.

**Review by gaborgsomogyi — APPROVED** [review]

File: external/avro/src/test/java/org/apache/spark/sql/avro/JavaAvroFunctionsSuite.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  private static void checkAnswer(Dataset<Row> actual, Dataset<Row> expected) {
+    String errorMessage = QueryTest$.MODULE$.checkAnswer(actual, expected.collectAsList());
+    if (errorMessage != null) {
+      Assert.fail(errorMessage);
+    }
+  }
+
+  @Test
+  public void testToAvroFromAvro() {
+      Dataset<Long> rangeDf = spark.range(10);
```
Comment (gaborgsomogyi) [inline external/avro/src/test/java/org/apache/spark/sql/avro/JavaAvroFunctionsSuite.java]: Sorry for nitpicking but the indent looks weird here.

**Review by gengliangwang — APPROVED** [review]
LGTM

**Review by viirya — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

Thanks, @gaborgsomogyi, @gengliangwang, @viirya

**Comment by cloud-fan** [comment]
a late LGTM

File: external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala [hunk 8]
```diff
@@ -34,30 +32,11 @@ package object avro {
...
-   * arbitrary result.
-   *
-   * @param data the binary column.
-   * @param jsonFormatSchema the avro schema in JSON string format.
-   * @param options options to control how the Avro record is parsed.
-   *
-   * @since 3.0.0
-   */
-  @Experimental
-  def from_avro(
```
Comment (zsxwing) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: @HyukjinKwon why did you remove this method directly rather than deprecating it?

Comment (cloud-fan) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: from the since annotation seems this was also added in 3.0 only?

Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: Yup.

Comment (zsxwing) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/package.scala]: Ah, yeah. I missed it.
